File: sdk/typescript/test/zip321.test.ts
import { describe, it, expect } from "vitest";
import {
  toBase64Url,
  decodeBase64Url,
  buildZcashUri,
  parseZip321Uri,
} from "../src/zip321.js";

describe("toBase64Url / decodeBase64Url", () => {
  it("round-trips plain text", () => {
    const text = "ZNS:CLAIM:alice:u1testaddr";
    expect(decodeBase64Url(toBase64Url(text))).toBe(text);
  });

  it("round-trips unicode", () => {
    const text = "hello world 🎉";
    expect(decodeBase64Url(toBase64Url(text))).toBe(text);
  });

  it("produces URL-safe characters", () => {
    const encoded = toBase64Url("test+/data==");
    expect(encoded).not.toMatch(/[+/=]/);
  });
});

describe("buildZcashUri", () => {
  it("builds address-only URI", () => {
    expect(buildZcashUri("u1addr")).toBe("zcash:u1addr");
  });

  it("builds URI with amount", () => {
    expect(buildZcashUri("u1addr", "0.25")).toBe("zcash:u1addr?amount=0.25");
  });

  it("builds URI with amount and memo", () => {
    const uri = buildZcashUri("u1addr", "1.5", "ZNS:CLAIM:alice:u1buyer");
    expect(uri).toMatch(/^zcash:u1addr\?amount=1\.5&memo=/);
  });

  it("accepts numeric amount", () => {
    expect(buildZcashUri("u1addr", 0.75)).toBe("zcash:u1addr?amount=0.75");
  });

  it("skips zero amount", () => {
    expect(buildZcashUri("u1addr", 0)).toBe("zcash:u1addr");
  });

  it("returns empty for empty address", () => {
    expect(buildZcashUri("")).toBe("");
  });
});

describe("parseZip321Uri", () => {
  it("parses a full URI", () => {
    const memo = "ZNS:CLAIM:alice:u1buyer";
    const uri = buildZcashUri("u1addr", "1.5", memo);
    const parts = parseZip321Uri(uri);
    expect(parts.address).toBe("u1addr");
    expect(parts.amount).toBe("1.5");
    expect(parts.memoDecoded).toBe(memo);
  });

  it("parses address-only URI", () => {
    const parts = parseZip321Uri("zcash:u1addr");
    expect(parts.address).toBe("u1addr");
    expect(parts.amount).toBe("");
    expect(parts.memoDecoded).toBe("");
  });
});

File: sdk/dart/test/zip321_test.dart
import 'package:test/test.dart';
import 'package:zcashname_sdk/zcashname_sdk.dart';

void main() {
  group('parseZip321Uri', () {
    test('parses address and amount', () {
      final r = parseZip321Uri('zcash:u1abc?amount=1.5&memo=abc123');
      expect(r.address, 'u1abc');
      expect(r.amount, '1.5');
    });

    test('decodes base64url memo', () {
      final r = parseZip321Uri('zcash:u1abc?memo=SGVsbG8');
      expect(r.memoDecoded, 'Hello');
    });

    test('returns empty strings for missing fields', () {
      final r = parseZip321Uri('zcash:u1abc');
      expect(r.amount, '');
      expect(r.memoRaw, '');
      expect(r.memoDecoded, '');
    });

    test('handles ZCASH: prefix case-insensitively', () {
      final r = parseZip321Uri('ZCASH:u1abc');
      expect(r.address, 'u1abc');
    });
  });

  group('buildZcashUri', () {
    test('returns base URI with no params when amount is null', () {
      expect(buildZcashUri('u1abc'), 'zcash:u1abc');
    });

    test('omits amount when 0', () {
      expect(buildZcashUri('u1abc', amountZats: 0), 'zcash:u1abc');
    });

    test('includes amount converted to ZEC', () {
      expect(buildZcashUri('u1abc', amountZats: 100000000),
          'zcash:u1abc?amount=1');
      expect(buildZcashUri('u1abc', amountZats: 10000001),
          'zcash:u1abc?amount=0.10000001');
      expect(buildZcashUri('u1abc', amountZats: 1000000),
          'zcash:u1abc?amount=0.01');
    });

    test('includes base64url-encoded memo', () {
      final uri = buildZcashUri('u1abc', memo: 'Hello');
      expect(uri.startsWith('zcash:u1abc?memo='), isTrue);
    });

    test('encodes memo without padding', () {
      // "Hello" base64-encoded with padding is "SGVsbG8=", without it's "SGVsbG8"
      final uri = buildZcashUri('u1abc', memo: 'Hello');
      final memoSegment = uri.split('memo=').last;
      expect(memoSegment.contains('='), isFalse);
    });
  });

  group('base64url helpers', () {
    test('toBase64Url/decodeBase64Url roundtrip', () {
      const input = 'Hello, ZNS world! 🦀';
      final encoded = toBase64Url(input);
      expect(encoded.contains('='), isFalse);
      expect(decodeBase64Url(encoded), input);
    });

    test('decodeBase64Url returns empty on bad input', () {
      expect(decodeBase64Url('@@@'), '');
    });
  });
}
